diffsims/sims/diffraction_simulation.py

```diff
@@ -44,35 +45,103 @@ class DiffractionSimulation:
 
     def __init__(
         self,
-        coordinates=None,
+        coordinates,
         indices=None,
         intensities=None,
-        calibration=1.0,
+        calibration=None,
         offset=(0.0, 0.0),
         with_direct_beam=False,
     ):
         """Initializes the DiffractionSimulation object with data values for
         the coordinates, indices, intensities, calibration and offset.
         """
-        self._coordinates = None
-        self.coordinates = coordinates
-        self.indices = indices
-        self._intensities = None
-        self.intensities = intensities
-        self._calibration = (1.0, 1.0)
+        if coordinates.ndim == 1:
+            coordinates = coordinates[None, :]
+        if indices is None:
+            indices = np.full((coordinates.shape[0], 3), np.nan)
+        if intensities is None:
+            intensities = np.full((coordinates.shape[0]), np.nan)
+        # check here whether shapes are all the same
+        if (
+            coordinates.shape[0] == indices.shape[0] == intensities.shape[0]
+            and coordinates.ndim == indices.ndim == 2
+            and intensities.ndim == 1
+        ):
+            self._coordinates = coordinates
+            self._indices = indices
+            self._intensities = intensities
+        else:
+            raise ValueError(
+                "Coordinate, intensity, and indices lists must be of the correct and matching shape."
+            )
         self.calibration = calibration
         self.offset = offset
         self.with_direct_beam = with_direct_beam
 
+    def __len__(self):
+        return self.coordinates.shape[0]
+
+    def __getitem__(self, sliced):
```

Comment:
Just for reference any numpy slice notation is valid, e.g. also `2:5`. However, numpy is very generous in the allowed slicing, for example passing in `None` can create new axes which would break the new object. I only allow slices that keep the array dimensions the same.